pkg/client/intermediate/intermediate.go

```diff
@@ -0,0 +1,139 @@
+package intermediate
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"sort"
+
+	"github.com/bitnami-labs/charts-syncer/pkg/client"
+
+	"github.com/juju/errors"
+
+	"github.com/bitnami-labs/charts-syncer/internal/utils"
+	"github.com/bitnami-labs/charts-syncer/pkg/client/types"
+	"helm.sh/helm/v3/pkg/chart"
+)
+
+var (
+	versionRe = regexp.MustCompile("(.*)-(\\d+\\.\\d+\\.\\d+)\\.bundle.tar")
+)
+
+// Repo allows to operate a chart repository.
+type Repo struct {
+	dir     string
+	entries map[string][]string
+}
+
+// NewIntermediateClient returns a ReadWriter object
+var NewIntermediateClient = func(intermediateBundlesPath string) (client.ReadWriter, error) {
+	return New(intermediateBundlesPath)
+}
+
+// New creates a Repo object from an api.Repo object.
+func New(dir string) (*Repo, error) {
+	d, err := filepath.Abs(dir)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err := os.MkdirAll(d, 0755); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Populate entries from directory
+	entries := make(map[string][]string)
+	matches, err := filepath.Glob(filepath.Join(d, "*.tar"))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, m := range matches {
+		filename := filepath.Base(m)
+		s := versionRe.FindStringSubmatch(filename)
+		entries[s[1]] = append(entries[s[1]], s[2])
+		sort.Strings(entries[s[0]])
+	}
+
+	return &Repo{dir: d, entries: entries}, nil
+}
+
+// List lists all chart names in a repo
+func (r *Repo) List() ([]string, error) {
+	var names []string
+	for name := range r.entries {
+		names = append(names, name)
+	}
+	return names, nil
+}
+
+// ListChartVersions lists all versions of a chart
+func (r *Repo) ListChartVersions(name string) ([]string, error) {
```

Comment:
following the pattern I mentioned above, this method would call underneath list and then extract the versions.

Reply:
+1